src/Microsoft.AspNetCore.Razor.Language/RazorProjectEngine.cs

```diff
@@ -47,10 +59,26 @@ public virtual RazorCodeDocument ProcessDesignTime(RazorProjectItem projectItem)
             return codeDocument;
         }
 
+        internal virtual RazorCodeDocument ProcessDesignTime(RazorSourceDocument source, IReadOnlyList<RazorSourceDocument> importSources, IReadOnlyList<TagHelperDescriptor> tagHelpers)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            var codeDocument = CreateCodeDocumentDesignTimeCore(source, importSources, tagHelpers);
+            ProcessCore(codeDocument);
+            return codeDocument;
+        }
+
         protected abstract RazorCodeDocument CreateCodeDocumentCore(RazorProjectItem projectItem);
 
+        internal abstract RazorCodeDocument CreateCodeDocumentCore(RazorSourceDocument source, IReadOnlyList<RazorSourceDocument> importSources, IReadOnlyList<TagHelperDescriptor> tagHelpers);
```

Comment:
This is a breaking change. You can't add a new abstract members

Reply:
👍 Will make it virtual